src/lib/userStorage/UserProfileStorage.js

```diff
@@ -0,0 +1,556 @@
+    const outputProfile = {}
+    await Promise.all(
+      Object.keys(profile).map(
+        async item =>
+          typeof profile[item]?.value === 'string' &&
+          (outputProfile[item] = {
+            ...profile[item],
+            value: await this.profiledb.decryptField(profile[item]?.value),
+          }),
+      ),
+    )
+    return outputProfile
+  }
+
+  /**
+   * encrypt decrypted profile
+   * @param {*} profile
+   */
+  async _encryptProfileFields(profile: { [key: string]: ProfileField }): Promise<any> {
+    let encryptProfile = {}
+    await Promise.all(
+      Object.keys(profile).map(async field =>
+        typeof profile[field]?.value === 'string'
+          ? (encryptProfile[field] = {
+              ...profile[field],
+              value: await this.profiledb.encryptField(profile[field]?.value),
+            })
+          : profile[field]?.value === null &&
+            (encryptProfile[field] = {
+              ...profile[field],
+            }),
+      ),
+    )
+    return encryptProfile
+  }
+
+  /**
+   * saves a complete profile to the underlying storage
+   * @param {*} profile
+   * @param update
+   */
+  async setProfile(profile: { [key: string]: string }, update: boolean = false): Promise<any> {
+    if (profile && !profile.validate) {
+      profile = getUserModel(profile)
+    }
+
+    const fields = Object.keys(profile).filter(prop => this.profileSettings[prop])
+    let { errors, isValid } = profile.validate(update)
+
+    if (!isValid) {
+      logger.warn(
+        'setProfile failed',
+        'Fields validation failed',
+        new Error('setProfile failed: Fields validation failed'),
+        { errors, category: ExceptionCategory.Human },
+      )
+
+      throw errors
+    }
+
+    const { avatar } = profile
+
+    if (!!avatar && isValidBase64Image(avatar)) {
+      const cids = await this._resizeAndStoreAvatars(avatar)
+
+      assign(profile, cids)
+    }
+
+    const fieldsToSave = fields.reduce(
+      (acc, currKey) => ({
+        ...acc,
+        [currKey]: {
+          value: profile[currKey],
+          display: profile[currKey],
+          privacy: this.getFieldPrivacy(currKey),
+        },
+      }),
+      {},
+    )
+    if (update) {
+      return this.setProfileFields(fieldsToSave)
+    }
+    return this.setNewProfileFields(fieldsToSave)
+  }
+
+  /**
+   * gets the current unencrypted field values in-memory
+   * @returns
+   */
+  getProfile(): { [key: string]: string } {
+    return Object.keys(this.profile).reduce((acc, currKey) => ({ ...acc, [currKey]: this.profile[currKey].value }), {})
+  }
+
+  /**
+   * updates profile fields in realm
+   * @param {*} fields
+   * @returns
+   */
+  async setProfileFields(fields: { [key: string]: ProfileField }): Promise<any> {
+    const encryptedFields = await this._encryptProfileFields(fields)
+    await this.profiledb.setProfileFields(encryptedFields)
```

Comment:
hmm can we merge profiledb.setProfileFields and setProfile?

Reply:
If u have on mind profile.setProfile then we need this option { upsert: true }in updateOne because of cases when profile not exists